### PR TITLE
LibWeb: Slottables, slots, slot assignments, shadow trees, and finally: `<details>`

### PR DESCRIPTION
**Pull request by trflynn89**
This implements the slot and slottable concepts from the DOM specs for use within the `<details>` element's shadow tree. These concepts are pretty beefy, but I've tried to break the implementation into understandable chunks.

https://github.com/SerenityOS/serenity/assets/5600524/73ed879a-b226-4005-aa7d-e3a26b9b734f

This also makes `console.group` work as intended in the JS console (though we can certainly circle back and style this a bit better):

https://github.com/SerenityOS/serenity/assets/5600524/d2f99866-e536-493e-acb8-c290a4a16ce3

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/ToggleEvent.idl [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#import <DOM/Event.idl>
+
+// https://html.spec.whatwg.org/multipage/interaction.html#toggleevent
+[Exposed=Window, UseNewAKString]
```
Comment (shannonbooth) [inline Userland/Libraries/LibWeb/HTML/ToggleEvent.idl:4]: Can remove UseNewAKString

**Comment by shannonbooth** [comment]
Nice on removing the `virtual` in `set_attribute` etc, that was in my git stash :P

File: Userland/Libraries/LibWeb/DOM/Slottable.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    // 4. If the open flag is set and shadow’s mode is not "open", then return null.
+    if (open_flag == OpenFlag::Set && shadow->mode() != Bindings::ShadowRootMode::Open)
+        return nullptr;
+
+    // 5. If shadow’s slot assignment is "manual", then return the slot in shadow’s descendants whose manually assigned
+    //    nodes contains slottable, if any; otherwise null.
+    if (shadow->slot_assignment() == Bindings::SlotAssignmentMode::Manual) {
+        JS::GCPtr<HTML::HTMLSlotElement> slot;
+
+        shadow->for_each_child_of_type<HTML::HTMLSlotElement>([&](auto& child) {
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/DOM/Slottable.cpp:79]: "descendants" here means the entire subtree
https://dom.spec.whatwg.org/#concept-tree-descendant

```suggestion
        shadow->for_each_in_subtree_of_type<HTML::HTMLSlotElement>([&](auto& child) {
```

File: Userland/Libraries/LibWeb/DOM/Slottable.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        });
+
+        return slot;
+    }
+
+    // 6. Return the first slot in tree order in shadow’s descendants whose name is slottable’s name, if any; otherwise null.
+    auto const& slottable_name = slottable.visit([](auto const& node) { return node->slottable_name(); });
+    JS::GCPtr<HTML::HTMLSlotElement> slot;
+
+    shadow->for_each_child_of_type<HTML::HTMLSlotElement>([&](auto& child) {
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/DOM/Slottable.cpp:93]: Ditto

**Review by Lubrsi — CHANGES_REQUESTED** [review]
Had a quick skim and noticed a couple issues, will have a closer look later :^)

File: Userland/Libraries/LibWeb/DOM/EventDispatcher.cpp [hunk 4]
```diff
@@ -251,11 +259,27 @@ bool EventDispatcher::dispatch(JS::NonnullGCPtr<EventTarget> target, Event& even
...
+            // 1. If slottable is non-null:
+            if (slottable != nullptr) {
+                // 1. Assert: parent is a slot.
+                VERIFY(is<HTML::HTMLSlotElement>(parent));
+
+                // 2. Set slottable to null.
+                slottable = nullptr;
+
+                // 3. If parent’s root is a shadow root whose mode is "closed", then set slot-in-closed-tree to true.
+                if (is<Node>(*parent)) {
```
Comment (Lubrsi) [inline Userland/Libraries/LibWeb/DOM/EventDispatcher.cpp:271]: This is always true, given the assert on step 1.

**Pull request merged**